### ISSUE TITLE
Improve OpenAI API compatibility

### ISSUE THREAD

**Issue by lvyufeng** [lvyufeng/PocketLLM#110]
## Problem

cpp_engine implements **only 2 OpenAI endpoints**:

```cpp
// cpp_engine/core/openai_server.cpp
server.Get("/v1/models", ...);
server.Post("/v1/chat/completions", ...);
```

Missing standard endpoints:
- `/v1/completions` - raw text completion (non-chat)
- `/v1/embeddings` - text embeddings
- `/health`, `/ready`, `/alive` - health checks beyond empty `/`

Missing features in `/v1/chat/completions`:
- `logprobs` parameter (return log probabilities)
- `n` parameter (return N completions per prompt)
- `stop` parameter (custom stop sequences, only uses EOS token now)
- `presence_penalty`, `frequency_penalty` (repetition penalties)
- `user` parameter (request attribution)

## Target Behavior (OpenAI/vLLM/SGLang)

Full OpenAI-compatible API surface:

### 1. /v1/completions

```bash
curl http://localhost:8000/v1/completions \
  -H "Content-Type: application/json" \
  -d '{
    "model": "qwen",
    "prompt": "Once upon a time",
    "max_tokens": 50,
    "temperature": 0.8,
    "stream": true
  }'
```

Difference from `/v1/chat/completions`:
- Takes raw string `prompt`, not `messages` array
- No chat template applied
- Useful for non-chat models or direct prompting

### 2. /v1/embeddings

```bash
curl http://localhost:8000/v1/embeddings \
  -H "Content-Type: application/json" \
  -d '{
    "model": "qwen",
    "input": ["Hello world", "How are you?"]
  }'

# Response:
{
  "object": "list",
  "data": [
    {"object": "embedding", "embedding": [0.123, -0.456, ...], "index": 0},
    {"object": "embedding", "embedding": [0.789, 0.012, ...], "index": 1}
  ],
  "model": "qwen",
  "usage": {"prompt_tokens": 8, "total_tokens": 8}
}
```

Requires:
- Extract final hidden state from model (before LM head)
- Return as float vector (typically dim=4096 for Qwen)

### 3. Health Check Endpoints

```bash
# Liveness: is process alive?
curl http://localhost:8000/health
{"status": "ok"}

# Readiness: is model loaded and ready to serve?
curl http://localhost:8000/ready
{"status": "ready", "model": "qwen-3.5"}

# Kubernetes-style
curl http://localhost:8000/alive
{"alive": true}
```

### 4. Extended /v1/chat/completions Parameters

```python
# logprobs: return log probabilities
{
  "messages": [...],
  "logprobs": true,
  "top_logprobs": 5  # Return top-5 token logprobs per position
}

# Response includes:
{
  "choices": [{
    "message": {"content": "Hello"},
    "logprobs": {
      "content": [
        {"token": "Hello", "logprob": -0.123, "top_logprobs": [...]},
        ...
      ]
    }
  }]
}

# n: multiple completions
{
  "messages": [...],
  "n": 3  # Return 3 different completions
}

# stop: custom stop sequences
{
  "messages": [...],
  "stop": ["\n\n", "END"]  # Stop at these strings, not just EOS
}

# Repetition penalties
{
  "messages": [...],
  "presence_penalty": 0.5,   # Penalize any repeated token
  "frequency_penalty": 0.8   # Penalize based on frequency
}
```

### Phase 1: /v1/completions

```cpp
// cpp_engine/core/openai_server.cpp

server.Post("/v1/completions", [&](const httplib::Request& req, httplib::Response& res) {
    json body = json::parse(req.body);

    std::string prompt = body["prompt"];  // Raw string, not messages
    int max_tokens = body.value("max_tokens", 256);
    bool stream = body.value("stream", false);

    // Tokenize directly without chat template
    auto prompt_ids = python_sidecar->tokenize(prompt, /*add_special_tokens=*/false);

    // ... same generation logic as chat/completions

    json response = {
        {"object", "text_completion"},
        {"choices", {{
            {"text", generated_text},
            {"index", 0},
            {"finish_reason", "length"}
        }}}
    };
    res.set_content(response.dump(), "application/json");
});
```

### Phase 2: /v1/embeddings

```cpp
// Requires engine API extension

// cpp_engine/include/persistent_engine.hpp
class PersistentEngine {
public:
    // NEW: extract hidden state instead of sampling
    std::vector<float> encode(const std::vector<int>& token_ids);
};

// Server endpoint
server.Post("/v1/embeddings", [&](const httplib::Request& req, httplib::Response& res) {
    json body = json::parse(req.body);

    std::vector<std::string> inputs;
    if (body["input"].is_string()) {
        inputs.push_back(body["input"]);
    } else {
        inputs = body["input"].get<std::vector<std::string>>();
    }

    json data = json::array();
    for (size_t i = 0; i < inputs.size(); ++i) {
        auto token_ids = python_sidecar->tokenize(inputs[i]);
        auto embedding = engine->encode(token_ids);

        data.push_back({
            {"object", "embedding"},
            {"embedding", embedding},
            {"index", i}
        });
    }

    json response = {
        {"object", "list"},
        {"data", data},
        {"model", "qwen"}
    };
    res.set_content(response.dump(), "application/json");
});
```

### Phase 3: Health Endpoints

```cpp
server.Get("/health", [](const httplib::Request&, httplib::Response& res) {
    json response = {{"status", "ok"}};
    res.set_content(response.dump(), "application/json");
});

server.Get("/ready", [&](const httplib::Request&, httplib::Response& res) {
    bool ready = (engine != nullptr);  // Model loaded?
    res.status = ready ? 200 : 503;
    json response = {
        {"status", ready ? "ready" : "not_ready"},
        {"model", ready ? "qwen-3.5" : ""}
    };
    res.set_content(response.dump(), "application/json");
});

server.Get("/alive", [](const httplib::Request&, httplib::Response& res) {
    json response = {{"alive", true}};
    res.set_content(response.dump(), "application/json");
});
```

### Phase 4: Extended Parameters (logprobs, n, stop, penalties)

```cpp
// cpp_engine/include/sampling.hpp (if it existed)
struct SamplingParams {
    float temperature = 0.0f;
    float top_p = 1.0f;
    int top_k = 20;
    bool greedy = false;

    // NEW fields
    bool return_logprobs = false;
    int top_logprobs = 0;
    std::vector<std::string> stop_sequences;
    float presence_penalty = 0.0f;
    float frequency_penalty = 0.0f;
};

// Requires engine changes to:
// 1. Track token counts for frequency_penalty
// 2. Apply penalties in sampling logits
// 3. Return logprobs alongside sampled token
// 4. Check stop_sequences after each token
```

## Success Criteria

```bash
# Test /v1/completions
curl -X POST http://localhost:8000/v1/completions \
  -d '{"prompt": "Hello", "max_tokens": 10}'
# Returns: {"choices": [{"text": "..."}]}

# Test /v1/embeddings
curl -X POST http://localhost:8000/v1/embeddings \
  -d '{"input": "Hello world"}'
# Returns: {"data": [{"embedding": [...]}]}

# Test health
curl http://localhost:8000/health
# Returns 200: {"status": "ok"}

# Test logprobs
curl -X POST http://localhost:8000/v1/chat/completions \
  -d '{"messages": [...], "logprobs": true}'
# Returns: {"choices": [{"logprobs": {...}}]}
```

## References

- OpenAI API spec: https://platform.openai.com/docs/api-reference
- vLLM OpenAI server: https://github.com/vllm-project/vllm/blob/main/vllm/entrypoints/openai/api_server.py
- SGLang OpenAI compatibility: https://github.com/sgl-project/sglang/blob/main/python/sglang/srt/openai_api_adapter.py

## Related Issues

- #105 Python library API (embeddings via `engine.encode()`)
- #108 Replace environment variables (penalties should be API parameters)

**Comment by lvyufeng** [comment]
## Progress Update

**Completed in PR #180:**
- ✅ Phase 1: POST /v1/completions endpoint (streaming and non-streaming)
- ✅ Phase 3: GET /ready and GET /alive health endpoints (GET /health already existed from #107)

**Remaining work:**
- ⏳ Phase 2: POST /v1/embeddings - requires new `InferenceEngine::encode()` API to extract hidden states
- ⏳ Phase 4: logprobs, n>1, presence_penalty, frequency_penalty - requires sampler infrastructure changes

Phase 2 and Phase 4 are blocked on deeper engine changes and should be tracked as separate follow-up issues.

**Comment by lvyufeng** [comment]
PR #180 已完成 Phase 1 (/v1/completions) 和 Phase 3 (health endpoints)。

剩余工作：
- Phase 2: /v1/embeddings (需要 InferenceEngine::encode() 方法)
- Phase 4: logprobs, presence_penalty, frequency_penalty, n>1 (需要采样器基础设施更改)

Phase 2 和 Phase 4 需要更深层次的引擎更改，建议拆分为独立的 issue。

**Comment by lvyufeng** [comment]
PR #180 完成了 Phase 1 和 Phase 3。

剩余工作：
- **Phase 2: /v1/embeddings** - 需要在 InferenceEngine 中添加 encode() 方法来提取隐藏状态
- **Phase 4: Extended Parameters** - 需要在采样器中添加 logprobs、presence_penalty、frequency_penalty、stop_sequences 支持

建议为 Phase 2 和 Phase 4 创建独立的 issue，因为它们需要引擎和采样器的深层更改。

**Comment by lvyufeng** [comment]
PR #180 已完成 Phase 1 (/v1/completions) 和 Phase 3 (health endpoints)。

Phase 2 (/v1/embeddings) 和 Phase 4 (logprobs, penalties, n>1, custom stop) 需要深层引擎改造，应该拆分为独立的 issue：

**Phase 2: /v1/embeddings**
- 需要在 InferenceEngine 中添加 encode() 方法
- 需要从 QwenEngine::Impl::run_chunk() 提取 hidden state
- 需要实现 mean-pooling 逻辑
- 需要处理 TP 下的 hidden state aggregation
- 需要添加 HTTP endpoint

**Phase 4: Extended parameters**
- logprobs/top_logprobs: 需要 sampler 返回 per-token log probabilities
- presence_penalty/frequency_penalty: 需要 token history tracking
- n > 1: 需要 request expansion 和 response aggregation
- custom stop sequences: 需要 tokenizer-aware multi-token matching

建议关闭此 issue（Phase 1+3 已完成），并为 Phase 2 和 Phase 4 分别创建新的 issue。
